instrumentation/grpc/src/main/java/brave/grpc/TracingClientInterceptor.java

```diff
@@ -127,38 +192,23 @@
 
     @Override public void onMessage(RespT message) {
       try (Scope scope = currentTraceContext.maybeScope(invocationContext)) {
+        Span span = spanRef.get(); // could be an error
+        SpanCustomizer customizer = span != null ? span.customizer() : NoopSpanCustomizer.INSTANCE;
+        messageProcessor.onMessageReceived(message, customizer);
         delegate().onMessage(message);
       }
     }
 
     @Override public void onClose(Status status, Metadata trailers) {
+      Throwable error = null;
       try (Scope scope = currentTraceContext.maybeScope(invocationContext)) {
         delegate().onClose(status, trailers);
-      }
-    }
-  }
-
-  final class TracingClientCallListener<RespT> extends SimpleForwardingClientCallListener<RespT> {
-    final Span span;
-
-    TracingClientCallListener(Listener<RespT> responseListener, Span span) {
-      super(responseListener);
-      this.span = span;
-    }
-
-    @Override public void onMessage(RespT message) {
-      try (Scope scope = currentTraceContext.maybeScope(span.context())) {
-        parser.onMessageReceived(message, span.customizer());
-        delegate().onMessage(message);
-      }
-    }
-
-    @Override public void onClose(Status status, Metadata trailers) {
-      try (Scope scope = currentTraceContext.maybeScope(span.context())) {
-        super.onClose(status, trailers);
-        parser.onClose(status, trailers, span.customizer());
+      } catch (Throwable e) {
+        error = e;
```

Comment:
Not sure but I think this would only happen if there is an error in the business logic on the client side when processing the closure, even if the request itself succeeded. So this might not warrant marking the span as error.

Reply:
agree. made a note in the rationale